processor/geoipprocessor/internal/provider/maxmindprovider/provider.go

```diff
@@ -0,0 +1,103 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package maxmind // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/geoipprocessor/internal/provider/maxmindprovider"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/oschwald/geoip2-golang"
+	"go.opentelemetry.io/otel/attribute"
+
+	conventions "github.com/open-telemetry/opentelemetry-collector-contrib/processor/geoipprocessor/internal/convention"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/geoipprocessor/internal/provider"
+)
+
+var (
+	// defaultLocale specifies English as the default Geolocation name, see https://dev.maxmind.com/geoip/docs/web-services/responses#languages
+	defaultLocale      = "en"
+	geoIP2CityDBType   = "GeoIP2-City"
+	geoLite2CityDBType = "GeoLite2-City"
+
+	errUnsupportedDB   = errors.New("unsupported geo IP database type")
+	errNoMetadataFound = errors.New("no geo IP metadata found")
+)
+
+type maxMindProvider struct {
+	geoReader *geoip2.Reader
+	// language/locale to be used in name retrieval
+	name string
+}
+
+var _ provider.GeoIPProvider = (*maxMindProvider)(nil)
+
+func newMaxMindProvider(cfg *Config) (*maxMindProvider, error) {
+	geoReader, err := geoip2.Open(cfg.GeoIPDatabasePath)
+	if err != nil {
+		return nil, fmt.Errorf("could not open geoip database: %w", err)
+	}
+
+	return &maxMindProvider{geoReader: geoReader, name: defaultLocale}, nil
+}
+
+// Location implements provider.GeoIPProvider for MaxMind. If a non City database type is used or no metadata is found in the database, an error will be returned.
+func (g *maxMindProvider) Location(_ context.Context, ipAddress net.IP) (attribute.Set, error) {
+	switch g.geoReader.Metadata().DatabaseType {
+	case geoIP2CityDBType, geoLite2CityDBType:
+		attrs, err := g.cityAttributes(ipAddress)
+		if err != nil {
+			return attribute.Set{}, err
+		} else if len(*attrs) == 0 {
+			return attribute.Set{}, errNoMetadataFound
+		}
+		return attribute.NewSet(*attrs...), nil
+	default:
+		return attribute.Set{}, fmt.Errorf("%w type: %s", errUnsupportedDB, g.geoReader.Metadata().DatabaseType)
+	}
+}
+
+// cityAttributes returns a list of key-values containing geographical metadata associated to the provided IP. The key names are populated using the internal geo IP conventions package. If the an invalid or nil IP is provided, an error is returned.
+func (g *maxMindProvider) cityAttributes(ipAddress net.IP) (*[]attribute.KeyValue, error) {
+	attributes := make([]attribute.KeyValue, 0, 11)
+
+	city, err := g.geoReader.City(ipAddress)
```

Comment:
nit: we could check `ipAddress.IsUnspecified` before making a call.

Reply:
Sounds good to me. I am wondering if it should be a provider check or at the processor level, it seems common enough to share it among all providers. Also, would it make sense to add the `ipAddress.IsLoopback` check too?

Reply:
common sounds good.
i think loopback address range can be useful. let's keep it supported for now

Reply:
Great! Check added in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33451/commits/8762af05d1b9a773b8f1b86fcf30ad5c170f3d57